### PR TITLE
Refactor `ClusterWideConfigurationService` to reduce duplication of configuration declarations

### PR DESCRIPTION
**Pull request by JackPGreen**
While trying to add a new configuration element to `ClusterWideConfigurationService`, I realised that the new element must be declared in multiple places:

1. `Map`s of `name` -> instances (e.g. `mapConfigs`)
2. `allConfigurations` - an array of every instance of (1)
3. `CONFIG_TO_VERSION` - a `Map` of handles configs and their introduction versions
4. `registerConfigLocally` - a method that, if applicable, adds the instance to (1)
5. The getters for (1) (e.g. `getMapConfigs()`)

I refactored the code to remove the need to update (1) & (2) and simplify (4).

Specifically:
- `allConfigurations` is now a `Map` of `Class` to (1), lazily created as required
- `registerConfigLocally` has had all of the common operations (if it's an `XConfig`, `putIfAbsent` to the `XConfig` `Map)  combined for all types
- The getters now query the map for the specified config class, rather than accessing the field directly
- Refactored `DynamicConfigVersionTest` to be a parameterized JUnit5 test

Pros:
- 15% less LOC
- need to update half as many places when adding new config types
- less repeated code = less chance of duplications getting out of sync

Cons:
- cognitive complexity increase
- the getters accessing via a `Map` will be slower than pure field access, but I hope we don't hit the dynamic config enough for that to be of any significance

### CODE REVIEW

File: hazelcast/src/main/java/com/hazelcast/internal/dynamicconfig/ClusterWideConfigurationService.java [hunk 1]
```diff
@@ -272,81 +252,43 @@ private IdentifiedDataSerializable cloneConfig(IdentifiedDataSerializable config
...
-                nodeEngine.getWanReplicationService().addWanReplicationConfig(config);
+        if ((newConfig instanceof MultiMapConfig) || (newConfig instanceof MapConfig)
+                || (newConfig instanceof CardinalityEstimatorConfig) || (newConfig instanceof RingbufferConfig)
+                || (newConfig instanceof ListConfig) || (newConfig instanceof SetConfig)
+                || (newConfig instanceof ReplicatedMapConfig) || (newConfig instanceof TopicConfig)
+                || (newConfig instanceof ExecutorConfig) || (newConfig instanceof DurableExecutorConfig)
+                || (newConfig instanceof ScheduledExecutorConfig) || (newConfig instanceof QueueConfig)
+                || (newConfig instanceof ReliableTopicConfig) || (newConfig instanceof CacheSimpleConfig)
+                || (newConfig instanceof FlakeIdGeneratorConfig) || (newConfig instanceof PNCounterConfig)
+                || (newConfig instanceof DataConnectionConfig) || (newConfig instanceof WanReplicationConfig)) {
```
Comment (JamesHazelcast) [inline hazelcast/src/main/java/com/hazelcast/internal/dynamicconfig/ClusterWideConfigurationService.java:266]: I'm not a fan of this huge conditional statement - I feel that the original layout was more understandable (albeit more LOC, but that's not really important over code readability). If we do keep this approach, I think it makes more sense to have a collection of supported configs (use version map?) that we use for checking instead.

File: hazelcast/src/test/java/com/hazelcast/internal/dynamicconfig/DynamicConfigVersionTest.java [hunk 2]
```diff
@@ -22,51 +22,46 @@
...
+    /** config classes not supported by dynamic data structure config */
+    private static Set<Class<?>> nonDynamicConfigClasses;
 
-    // config classes not supported by dynamic data structure config
-    private static final Set<Class<?>> NON_DYNAMIC_CONFIG_CLASSES;
+    @BeforeAll
+    public static void setUp() {
+        nonDynamicConfigClasses = Set.of(WanReplicationConfig.class, SplitBrainProtectionConfig.class, ListenerConfig.class,
+                DeviceConfig.class, DynamicConfigurationConfig.class);
+    }
```
Comment (JamesHazelcast) [inline hazelcast/src/test/java/com/hazelcast/internal/dynamicconfig/DynamicConfigVersionTest.java:49]: I feel this was better left as a statically initialized value (could still leverage `Set.of()`)

**Review by JamesHazelcast — COMMENTED** [review]
A couple of direct comments, but in general I'm not sure if I agree with these changes - they make the code much harder to understand at a glance imo, and introduces additional performance overhead for the sake of fewer LOC. I agree there's likely not very much performance impact in regards to dynamic configuration, but I know the lookup/find methods are called in our places too, so we should make sure none of this code is on a hot path before merging at minimum.

That being said, I do appreciate the reduction in work for adding new structures that this PR brings, but I'm just not sure if the trade-off is worth it (especially since we don't add to this structure very often).

File: hazelcast/src/test/java/com/hazelcast/internal/dynamicconfig/DynamicConfigVersionTest.java (hunk 2 above)
Comment (JackPGreen) [reply hazelcast/src/test/java/com/hazelcast/internal/dynamicconfig/DynamicConfigVersionTest.java:49]: > I feel this was better left as a statically initialized value (could still leverage `Set.of()`)

Can I ask why? My understanding of JUnit is that you want to ensure that everything is initialized via a `@Before`, both for testabilities sake, but also for performance - something about allowing JUnit to decide when they are initialized.

Comment (JamesHazelcast) [reply hazelcast/src/test/java/com/hazelcast/internal/dynamicconfig/DynamicConfigVersionTest.java:49]: That makes sense for non-static variables, but in this case the `Set` contents are never going to change and they're always needed for the test, so initialization order is not going to change anything really. In terms of performance, having it statically configured will give the compiler more room to optimize, whereas embedding within a method does not (on top of the minute performance overhead of the method calls).

Comment (JamesHazelcast) [reply hazelcast/src/test/java/com/hazelcast/internal/dynamicconfig/DynamicConfigVersionTest.java:49]: I also just noticed `WanReplicationConfig` is actually dynamic now, so this list needs updating (I believe @gbarnett-hz added dynamic support previously)

**Pull request closed without merging**